File: 014/BinarySearch.py
class Solution:
    # @param nums: The integer array
    # @param target: Target number to find
    # @return the first position of target in nums, position start from 0 
    def binarySearch(self, nums, target):
        # write your code here
        start = 0
        end = len(nums) - 1
        while start <= end:
            middle = (start + end) / 2
            if target == nums[middle]:
                for x in range(start, middle):
                    if nums[x] == target:
                        return x
                return middle
            elif target < nums[middle]:
                end = middle - 1
            else:
                start = middle + 1
        return -1
        

File: 014 二分查找/BinarySearch.cpp
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
	int binarySearch(vector<int> &array, int target) {
		// write your code here
		int start = 0;
		int end = array.size();
		int mid = 0;
		int index = -1;
		while (start <= end) 
		{
			mid = (start + end) / 2;
			if (target == array.at(mid)) {
				index = mid;
				while ((mid-1) >= 0 && array.at(mid-1) == target)
				{
					mid = mid - 1;
					index = mid;
				}
				return index;
			}
			else if (target < array.at(mid))
			{
				end = mid - 1;
			}
			else
			{
				start = mid + 1;
			}
		}
		return index;
	}
};

int main() {
	Solution s;
	vector<int> v;
	v.push_back(1);
	v.push_back(2);
	v.push_back(3);
	v.push_back(3);
	v.push_back(4);
	v.push_back(5);
	v.push_back(10);
	cout << s.binarySearch(v, 1) << "\n";
	// cout << v.at(4) << "\n";
}

